### PR TITLE
Update to dry-system-0.12

### PR DESCRIPTION
**Pull request by arielvalentin**
Update `dry-system` to 0.12.x to add support for memoization. 

Drop Support for Ruby < 2.4.

### CODE REVIEW

File: lib/dry/system/rails/version.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@
 module Dry
   module System
     module Rails
-      VERSION = '0.2.0'
+      VERSION = '0.2.1.pre'
```
Comment (solnic) [inline lib/dry/system/rails/version.rb]: Why not `0.3.0`?

**Comment by solnic** [comment]
We need to drop support for ruby `< 2.4` too

File: lib/dry/system/rails/version.rb (hunk 1 above)
Comment (arielvalentin) [reply lib/dry/system/rails/version.rb]: Good question. I don’t know what the release process looks like for the other gems in the eco system. 

Is that documented anywhere?

Comment (solnic) [reply lib/dry/system/rails/version.rb]: We just follow semver. Gems below 1.0 don’t need pre-release suffixes as they are beta/unstable.

It should be documented. I was actually supposed to do that 😅

Comment (arielvalentin) [reply lib/dry/system/rails/version.rb]: OK no worries.

**Pull request merged**